### ISSUE TITLE
Partial Refunds Failing - Stripe Gateway Add-on

### ISSUE THREAD

**Issue by iversoncreative** [ingenesis/shopp#3187]
**Description**
Issuing partial refunds from within Shopp's order management are failing for the Stripe gateway add-on. *I have also created a support ticket within Shopp support but want to tap the GitHub community as well.*

**Background**
The merchant offers a 10% discount code and wants to retroactively refund 10% of the order amount for users who do not enter the discount code at checkout. 

**Steps**
Once the order has been charged, we click the Refund button, select a reason for the refund, enter the amount, then click the Process Refund button. We then get a generic refund failed error as seen below. We are not getting any other errors from within Shopp.

**Error Screenshot**
![screen shot 2015-01-02 at 8 50 41 am](https://cloud.githubusercontent.com/assets/3998792/5597634/fb63cb9a-9262-11e4-8a40-98dce2cf72e1.png)

**Versions**
WordPress 4.0.1
Shopp 1.3.7
Stripe 1.1.1dev (dev version provided by Shopp support team)
PHP 5.3
MySQL 5.5.32

**Shopp Log**
id: XXX
object: "charge"
created: 1418146533
livemode: true
paid: true
amount: 3000
currency: "usd"
refunded: false
captured: true
card:
    id: XXX
    object: "card"
    last4: "XXX"
    brand: "XXX"
    funding: "debit"
    exp_month: XXX
    exp_year: XXX
    fingerprint: "XXX"
    country: "US"
    name: "XXX"
    address_line1: "XXX"
    address_line2: ""
    address_city: null
    address_state: "XXX"
    address_zip: "XXX"
    address_country: "XXX"
    cvc_check: "pass"
    address_line1_check: "pass"
    address_zip_check: "pass"
    dynamic_last4: null
    customer: null
balance_transaction: "XXX"
failure_message: null
failure_code: null
amount_refunded: 2295
customer: null
invoice: null
description: "XXX"
dispute: null
metadata:
statement_description: null
fraud_details:
receipt_email: null
receipt_number: null
shipping: null
refunds:
    object: "list"
    total_count: 5
    has_more: false
    url: "/v1/charges/XXX/refunds"
    data:
            id: XXX
            amount: 459
            currency: "usd"
            created: 1418854543
            object: "refund"
            balance_transaction: "XXX"
            metadata:
            charge: XXX
            receipt_number: null
            reason: null
            id: XXX
            amount: 459
            currency: "usd"
            created: 1418854421
            object: "refund"
            balance_transaction: "XXX"
            metadata:
            charge: XXX
            receipt_number: null
            reason: null
            id: XXX
            amount: 459
            currency: "usd"
            created: 1418854261
            object: "refund"
            balance_transaction: "XXX"
            metadata:
            charge: XXX
            receipt_number: null
            reason: null
            id: XXX
            amount: 459
            currency: "usd"
            created: 1418854147
            object: "refund"
            balance_transaction: "XXX"
            metadata:
            charge: XXX
            receipt_number: null
            reason: null
            id: XXX
            amount: 459
            currency: "usd"
            created: 1418779241
            object: "refund"
            balance_transaction: "XXX"
            metadata:
            charge: XXX
            receipt_number: null
            reason: null

**Comment by iversoncreative** [comment]
Hey, guys. Any progress on this issue? Anything I can help with in resolving it?

**Comment by jonathandavis** [comment]
I don't suppose you have the actual response from Stripe logged do you? I'll see if I can get this resolved today if at all possible.

**Comment by iversoncreative** [comment]
Where would one find that response logged? If you're referring to the Log within Shopp > System > Log, I turned on All Errors in System > Advanced > Logging and am copying/pasting the results below after attempting a partial refund, though I don't see anything regarding Stripe or refunds. If there is somewhere else you'd like me to look I'll do it.

2015-01-21 15:10:40 - PHP NOTICE: Undefined index: QUERY_STRING ShoppErrors->php, ShoppStorefront->dashboard, call_user_func_array, do_action_ref_array, WP->main, wp, require [Storefront.php, line 925]

2015-01-21 15:10:42 - PHP NOTICE: Undefined index: QUERY_STRING ShoppErrors->php, ShoppStorefront->dashboard, call_user_func_array, do_action_ref_array, WP->main, wp, require [Storefront.php, line 925]

2015-01-21 15:10:48 - PHP NOTICE: Undefined index: QUERY_STRING ShoppErrors->php, ShoppStorefront->dashboard, call_user_func_array, do_action_ref_array, WP->main, wp, require [Storefront.php, line 925]

2015-01-21 15:10:50 - PHP NOTICE: Undefined index: QUERY_STRING ShoppErrors->php, ShoppStorefront->dashboard, call_user_func_array, do_action_ref_array, WP->main, wp, require [Storefront.php, line 925]

2015-01-21 15:11:07 - PHP NOTICE: Undefined index: NOTRACKING ShoppErrors->php, ShippedOrderEventRenderer->carrier, ShippedOrderEventRenderer->carrier_name, ShippedOrderEventRenderer->details, manage_meta_box, call_user_func, ShoppUI->metabox, call_user_func, do_meta_boxes, include, ShoppAdminService->manager, ShoppAdminService->admin, ShoppFlow->admin, call_user_func_array, do_action [events.php, line 434]

2015-01-21 15:11:07 - PHP WARNING: array_combine(): Both parameters should have at least 1 element ShoppErrors->php, array_combine, OrderEventRenderer->load, OrderEventRenderer->__construct, OrderEventRenderer->renderer, OrderEventRenderer->display, call_user_func_array, apply_filters, history_meta_box, call_user_func, ShoppUI->metabox, call_user_func, do_meta_boxes, include, ShoppAdminService->manager, ShoppAdminService->admin, ShoppFlow->admin, call_user_func_array, do_action [events.php, line 52]

2015-01-21 15:11:07 - PHP NOTICE: Undefined index: NOTRACKING ShoppErrors->php, ShippedOrderEventRenderer->carrier, ShippedOrderEventRenderer->carrier_name, ShippedOrderEventRenderer->details, OrderEventRenderer->content, OrderEventRenderer->render, OrderEventRenderer->display, call_user_func_array, apply_filters, history_meta_box, call_user_func, ShoppUI->metabox, call_user_func, do_meta_boxes, include, ShoppAdminService->manager, ShoppAdminService->admin, ShoppFlow->admin, call_user_func_array, do_action [events.php, line 434]

2015-01-21 15:11:12 - PHP NOTICE: Undefined index: QUERY_STRING ShoppErrors->php, ShoppStorefront->dashboard, call_user_func_array, do_action_ref_array, WP->main, wp, require [Storefront.php, line 925]

2015-01-21 15:11:13 - PHP NOTICE: Undefined index: QUERY_STRING ShoppErrors->php, ShoppStorefront->dashboard, call_user_func_array, do_action_ref_array, WP->main, wp, require [Storefront.php, line 925]

2015-01-21 15:11:41 - PHP NOTICE: Undefined index: NOTRACKING ShoppErrors->php, ShippedOrderEventRenderer->carrier, ShippedOrderEventRenderer->carrier_name, ShippedOrderEventRenderer->details, manage_meta_box, call_user_func, ShoppUI->metabox, call_user_func, do_meta_boxes, include, ShoppAdminService->manager, ShoppAdminService->admin, ShoppFlow->admin, call_user_func_array, do_action [events.php, line 434]

2015-01-21 15:11:41 - PHP WARNING: array_combine(): Both parameters should have at least 1 element ShoppErrors->php, array_combine, OrderEventRenderer->load, OrderEventRenderer->__construct, OrderEventRenderer->renderer, OrderEventRenderer->display, call_user_func_array, apply_filters, history_meta_box, call_user_func, ShoppUI->metabox, call_user_func, do_meta_boxes, include, ShoppAdminService->manager, ShoppAdminService->admin, ShoppFlow->admin, call_user_func_array, do_action [events.php, line 52]

2015-01-21 15:11:41 - PHP NOTICE: Undefined index: NOTRACKING ShoppErrors->php, ShippedOrderEventRenderer->carrier, ShippedOrderEventRenderer->carrier_name, ShippedOrderEventRenderer->details, OrderEventRenderer->content, OrderEventRenderer->render, OrderEventRenderer->display, call_user_func_array, apply_filters, history_meta_box, call_user_func, ShoppUI->metabox, call_user_func, do_meta_boxes, include, ShoppAdminService->manager, ShoppAdminService->admin, ShoppFlow->admin, call_user_func_array, do_action [events.php, line 434]

2015-01-21 15:11:46 - PHP NOTICE: Undefined index: QUERY_STRING ShoppErrors->php, ShoppStorefront->dashboard, call_user_func_array, do_action_ref_array, WP->main, wp, require [Storefront.php, line 925]

**Comment by dxdc** [comment]
I can confirm this bug as well. Here is the Stripe output. Thanks Jonathan for working on a fix!

Event Data

object:
id: ch_15UUna2ACRl0JvZRfw8UXwnx
object: "charge"
created: 1423539294
livemode: false
paid: true
amount: 1691
currency: "usd"
refunded: false
captured: true
card:
id: card_15UUna2ACRl0JvZRD0WiTayT
object: "card"
last4: "4242"
brand: "Visa"
funding: "credit"
exp_month: 9
exp_year: 2020
fingerprint: "4Y2x9CjfZxYc9KL1"
country: "US"
name: "Donny"
address_line1: "1950 plain street"
address_line2: ""
address_city: null
address_state: "MD"
address_zip: "20901"
address_country: "US"
cvc_check: "pass"
address_line1_check: "pass"
address_zip_check: "pass"
dynamic_last4: null
customer: null
balance_transaction: "txn_15UUna2ACRl0JvZRM8vvWbWd"
failure_message: null
failure_code: null
amount_refunded: 191
customer: null
invoice: null
description: "element26@gmail.com"
dispute: null
metadata:
statement_descriptor: null
fraud_details:
receipt_email: null
receipt_number: null
shipping: null
refunds:
object: "list"
total_count: 1
has_more: false
url: "/v1/charges/ch_15UUna2ACRl0JvZRfw8UXwnx/refunds"
data:
id: re_15UWme2ACRl0JvZR22lF9arU
amount: 191
currency: "usd"
created: 1423546924
object: "refund"
balance_transaction: "txn_15UWme2ACRl0JvZRJP5p6a4C"
metadata:
charge: ch_15UUna2ACRl0JvZRfw8UXwnx
receipt_number: null
reason: null
previous_attributes:
amount_refunded: 0
refunds:
total_count: 0
data:

**Comment by jonathandavis** [comment]
A fix is ready, but is waiting on a final issue before release. It could land as soon as today or as late as the end of this week.

**Comment by dxdc** [comment]
thanks Jonathan! does this fix include an update to tokenize the card number per Stripe? That would really help me out.

**Comment by iversoncreative** [comment]
Thanks, @jonathandavis! Client will be happy to hear it. Much appreciated.

**Comment by dxdc** [comment]
The origin of this problem is this line in the Stripe.php, under public function refund {

		if ( ! $response || is_a($response,'ShoppError') || ! $response->refunded ) {

Because $response->refunded is false for partial refunds, this will always fail. I was able to fix this with the following modification:

		if ( ! $response || is_a($response,'ShoppError') || ! $response->amount_refunded ) {

Will be curious to see how Jonathan updates the Stripe code.

**Comment by jonathandavis** [comment]
This is resolved in the add-on repo. Release due soon. Another issue related to duplicate capture requests in Stripe is blocking release.

**Comment by jacksonwp** [comment]
Is this bug still present in the published version of the Stripe gateway? Have a customer complaining again about partial refunds in Stripe not working.

**Comment by sierrab123** [comment]
Just looked at the commits and can see that the fix went into 1.1.1

**Comment by jacksonwp** [comment]
Thanks, is Stripe supposed to update automatically when Shopp Support is active? Where can I view the add-on repro for reference?

**Comment by sierrab123** [comment]
Stripe does not auto update, is your support current for Stripe? Sadly you can't view the repo, it's behind closed doors. There has only been one commit since 1.1.1 was released and that is not do with this issue

**Comment by jacksonwp** [comment]
Looks like my support lapsed, don't ever get renewal notices for add-ons?

You might want to update this page: https://shopplugin.net/docs/support-updates/updates/

> Any officially supported purchased add-ons from shopplugin.net will automatically be updated with core plugin updates.
